python/metatensor-operations/metatensor/operations/add.py

```diff
@@ -106,6 +106,15 @@ def _add_block_block(block_1: TensorBlock, block_2: TensorBlock) -> TensorBlock:
     for parameter, gradient_1 in block_1.gradients():
         gradient_2 = block_2.gradient(parameter)
 
+        if gradient_1.values.shape != gradient_2.values.shape:
+            raise ValueError(
+                "The two gradient blocks must have the same shape. "
+                "Different sparsity patterns along the samples dimension "
+                "are not supported."
+            )
+        if gradient_1.samples != gradient_2.samples:
+            raise ValueError("The two gradient blocks must have the same samples")
+
```

Comment:
Both of these conditions are already checked in the main function with `_check_same_gradients_raise`, so instead of duplicating checks we can have an assert here (sorry, I should have caught that earlier)

```suggestion
        assert gradient_1.values.shape == gradient_2.values.shape
```

Reply:
And similarly for the other functions!

Reply:
Ahh my bad, I should have noticed as well